### PR TITLE
[compiler-v2] Enum types frontend

### PR DESCRIPTION
**Pull request by wrwg**
## Description

This implements the frontend (parser, context checker, and model representation) of struct variants aka 'enum types', including the `enum` type declaration, as well as the `match` expression. Some code examples [are in this test](https://github.com/aptos-labs/aptos-core/blob/wrwg/enum/third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move).

The internal design principle is that enum types are not a new kind of types, but just a variation of the existing struct types, which now have multiple variants. This allows the implementation to have a relative small footprint. The same principle is likely later to be applied for the VM implementation.

Many of the existing code for dealing with struct pack and unpack can be reused for enum types. Some changes in the expansion phase and the model builder have to be applied, as a struct name can now be accompanied by an optional variant name. The model builder has also been slightly refactored and cleaned up to deal with the new features.

The implementation does not yet check for exhaustive matches and similar conditions. Those checks are naturally done later by the code which compiles matches to bytecode.

There are some general changes to baseline files unrelated to the functionality of this PR but resulting from refining the Model dump representation, as well as minor bug fixes. For the tests relevant to this PR, see `test/checking/variants/`.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Existing and new tests

### CODE REVIEW

**Review by vineethk — COMMENTED** [review]
Looks good!

Have some minor comments and some questions.

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 1]
```diff
@@ -3724,40 +3935,118 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
...
+            .into_iter()
+            .sorted_by_key(|(i, _)| *i)
+            .map(|(_, value)| value)
+            .collect_vec();
+
+        let struct_ty = struct_id.to_type();
+        let struct_ty = self.check_type(loc, &struct_ty, expected_type, context);
+        let mut field_args = args.into_iter().map(|e| e.into_exp()).collect_vec();
+        if variant.is_none() && field_args.is_empty() {
+            // The move compiler inserts a dummy field with the value of false
```
Comment (rahxephon89) [inline third_party/move/move-model/src/builder/exp_builder.rs:4030]: I am wondering whether zero-variant enum can be allowed?

File: third_party/move/move-compiler-v2/tests/checking/typing/recursive_structs_malformed.exp [hunk 2]
```diff
@@ -16,7 +16,7 @@ error: reference type `&Bar` is not allowed as a field type
   │
   = required by declaration of field `f`
 
-error: tuple type `(&Baz, &mut Baz)` is not allowed as a type argument
+error: tuple type `(&Baz, &mut Baz)` is not allowed as a type instantiation
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/checking/typing/recursive_structs_malformed.exp:19]: The docs refer to this as a "type argument" or "type parameter".  See https://aptos.dev/move/book/generics/.

File: third_party/move/evm/move-to-yul/tests/dispatcher_testsuite.rs [hunk 3]
```diff
@@ -83,6 +87,11 @@ fn compile_yul_to_bytecode_bytes(filename: &str) -> Result<Vec<u8>> {
         flags,
         &known_attributes,
     )?;
+    if env.has_errors() {
+        let mut error_writer = StandardStream::stderr(ColorChoice::Auto);
```
Comment (brmataptos) [inline third_party/move/evm/move-to-yul/tests/dispatcher_testsuite.rs:91]: I feel like someone told me that tests shouldn't output in color.

File: third_party/move/move-compiler/src/expansion/translate.rs [hunk 4]
```diff
@@ -1649,6 +1690,10 @@ fn access_specifier_list_as_acquires(
                             && check_wildcard(context, second)
                             && check_wildcard(context, third)
                     },
+                    NameAccessChain_::Four(..) => {
```
Comment (brmataptos) [inline third_party/move/move-compiler/src/expansion/translate.rs:1693]: Can we get a test for this code?  Or do we have one already?

File: third_party/move/move-model/src/model.rs [hunk 5]
```diff
@@ -3376,7 +3456,24 @@ impl<'env> StructEnv<'env> {
...
-    /// Return the number of fields in the struct.
+    /// Get fields of a particular variant.
+    pub fn get_fields_of_variant(
+        &'env self,
+        variant: Symbol,
+    ) -> impl Iterator<Item = FieldEnv<'env>> {
+        self.data
+            .field_data
+            .values()
+            .filter(|data| data.variant == Some(variant))
```
Comment (brmataptos) [inline third_party/move/move-model/src/model.rs:3467]: This may be very expensive.  Add a warning to the function comment.

File: third_party/move/move-model/src/model.rs [hunk 6]
```diff
@@ -2310,15 +2322,30 @@ impl GlobalEnv {
...
-                        "{}: {},",
-                        fld.get_name().display(spool),
-                        fld.get_type().display(tctx)
-                    );
+                if str.has_variants() {
+                    emitln!(writer, "enum {} {{", str.get_name().display(spool));
+                    writer.indent();
+                    for variant in str.get_variants() {
+                        emit!(writer, "{}", variant.display(spool));
+                        let fields = str.get_fields_of_variant(variant).collect_vec();
```
Comment (brmataptos) [inline third_party/move/move-model/src/model.rs:2330]: Note that `get_fields_of_variant()` is somewhat expensive, as it will iterate over`field_data` for all variants each time you call it.  You may want to define a more targeted helper function like `get_variants_with_fields()` which can pull data for all variants in one pass.

File: third_party/move/move-model/src/lib.rs [hunk 7]
```diff
@@ -111,7 +111,14 @@ pub fn run_model_builder_in_compiler_mode(
         },
         Flags::model_compilation()
             .set_skip_attribute_checks(skip_attribute_checks)
-            .set_keep_testing_functions(compile_test_code),
+            .set_keep_testing_functions(compile_test_code)
+            .set_flavor(
+                if language_version != LanguageVersion::V1 {
```
Comment (brmataptos) [inline third_party/move/move-model/src/lib.rs:116]: Better if we can move this to a separate PR.

File: third_party/move/move-compiler/src/expansion/translate.rs (hunk 4 above)
Comment (rahxephon89) [reply third_party/move/move-compiler/src/expansion/translate.rs:1693]: I found one here: third_party/move/move-compiler-v2/tests/checking/variants/variants_check_err.move

File: third_party/move/move-compiler/src/parser/ast.rs [hunk 8]
```diff
@@ -1274,13 +1301,13 @@ impl AstDebug for StructDefinition {
...
 
-        if let StructFields::Native(_) = fields {
+        if let StructLayout::Native(_) = fields {
             w.write("native ");
         }
 
         w.write(&format!("struct {}", name));
         type_parameters.ast_debug(w);
-        if let StructFields::Defined(fields) = fields {
+        if let StructLayout::Singleton(fields) = fields {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/ast.rs:1310]: Ahh, those ast debug printers, I don't know of any use case, and they are already incomplete since long (e.g. abilities and attributes). Added a case now but it just prints variants are NYI for printing.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 9]
```diff
@@ -700,15 +730,19 @@ fn parse_bind(context: &mut Context) -> Result<Bind, Box<Diagnostic>> {
...
+    let ty = parse_name_access_chain(context, false, || "a variable or struct or variant name")?;
     let ty_args = parse_optional_type_args(context)?;
-    let args = parse_comma_list(
-        context,
-        Tok::LBrace,
-        Tok::RBrace,
-        parse_bind_field,
-        "a field binding",
-    )?;
+    let args = if !context.env.is_move_2() || context.tokens.peek() == Tok::LBrace {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:735]: Please remember for language design discussions.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 10]
```diff
@@ -1413,6 +1456,55 @@ fn parse_for_loop(context: &mut Context) -> Result<(Exp, bool), Box<Diagnostic>>
...
+    consume_token(context.tokens, Tok::RBrace)?;
+    Ok(spanned(
+        context.tokens.file_hash(),
+        start_loc,
+        context.tokens.previous_end_loc(),
+        Exp_::Match(Box::new(exp), arms),
+    ))
+}
+
+fn parse_match_arms(
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:1478]: The `MatchArm` syntax is documented above, I find it more adequate together with parse_match_exp, as this is just a helper function here

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 11]
```diff
@@ -2528,20 +2630,37 @@ fn parse_struct_decl(
...
     };
 
-    let fields = match native {
+    let layout = match native {
         Some(loc) => {
             consume_token(context.tokens, Tok::Semicolon)?;
-            StructFields::Native(loc)
+            StructLayout::Native(loc)
         },
         _ => {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:2638]: Done

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 12]
```diff
@@ -2867,8 +3023,13 @@ fn parse_module(
                             attributes, start_loc, modifiers, context,
                         )?),
                         Tok::Struct => ModuleMember::Struct(parse_struct_decl(
-                            attributes, start_loc, modifiers, context,
+                            false, attributes, start_loc, modifiers, context,
                         )?),
+                        Tok::Identifier if context.tokens.content() == "enum" => {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:3028]: I know only of this use case. We do it similarly already over this file for other soft keywords.

File: third_party/move/move-compiler/tests/move_check/expansion/use_nested_self_as_invalid.exp [hunk 13]
```diff
@@ -4,15 +4,18 @@ warning[W09001]: unused alias
...
+   │                   │
+   │                   Invalid field containing 'X' in struct 'X'.
+   │                   Using this struct creates a cycle: 'X' contains 'X'
 
 error[E03002]: unbound module
    ┌─ tests/move_check/expansion/use_nested_self_as_invalid.move:12:9
    │
 12 │         X::foo();
-   │         ^ Unbound module alias 'X'
+   │         ^ Unbound module or type alias 'X'
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/tests/move_check/expansion/use_nested_self_as_invalid.exp:20]: Perhaps, but the new semantics is like it us, it can be either of those cases now.

File: third_party/move/move-model/src/ast.rs [hunk 14]
```diff
@@ -1674,7 +1706,13 @@ pub enum Pattern {
     Var(NodeId, Symbol),
     Wildcard(NodeId),
     Tuple(NodeId, Vec<Pattern>),
-    Struct(NodeId, QualifiedInstId<StructId>, Vec<Pattern>),
+    // Struct(_, struct_id, optional_variant, patterns)
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/ast.rs:1709]: That's actually documentation. Move it a bit to make clearer.

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 15]
```diff
@@ -3724,40 +3935,118 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
...
+            &self.to_loc(&maccess.loc),
+            true,
+            struct_name.symbol,
+            &struct_entry.type_params,
+            generics,
+        )?;
+
+        // If this is a struct variant, check whether it exists.
+        if let Some(v) = variant {
+            if !self.check_variant_declared(&struct_name, &struct_entry, &struct_name_loc, v) {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/builder/exp_builder.rs:3958]: Noted

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-model/src/builder/exp_builder.rs:4030]: It is right now, there is no particular technical or semantic reason to forbid it. We can come back if we want to ban it.

File: third_party/move/move-model/src/lib.rs (hunk 7 above)
Comment (wrwg) [reply third_party/move/move-model/src/lib.rs:116]: I need it because the parser wasn't able to reflect yet over the language version. Now he can, over the flavor.

File: third_party/move/move-model/src/model.rs (hunk 6 above)
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:2330]: I think this is marginal, we are talking in the majority of cases about like 4 fields or so. Let us not make the code unnecessary complex because of such details.

File: third_party/move/move-model/src/model.rs (hunk 5 above)
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:3467]: Nothing particular expensive here, iterating over a list with a few elements.

**Review by wrwg — COMMENTED** [review]
Thanks for the reviews, PTAL

**Review by rahxephon89 — APPROVED** [review]

File: third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        match (self) {
+            Color::RGB{red, green, blue} => { red + green + blue > 0 },
+            Color::Red => true,
+            Color::Blue => false,
+        }
+    }
+
+    fun t1_uses_block_no_comma(self: Color): bool {
+        match (self) {
+            Color::RGB{red, green, blue} => { red + green + blue > 0 }
```
[the remark this answers is not in the document]
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move:42]: I see.  Well, the world is full of bad ideas.  Move may as well be, too.

File: third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    use 0x815::m;
+
+    enum Color {
+        RGB{red: u64, green: u64, blue: u64},
+        Red,
+        Blue,
+    }
+
+    enum ColorUsesBlockNoComma {
+        RGB{red: u64, green: u64, blue: u64}
```
[the remark this answers is not in the document]
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move:11]: If you don't have a language spec for your feature you are implementing, it would be very useful to provide references to related language specs you are emulating.  I was unfamiliar with this Rust feature.

File: third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move:42]: Whatever.

File: third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move (hunk 17 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move:11]: There is a design doc and there was a design review. The doc states, and it was discussed in the review, that we are orienting towards Rust.

Having said so, I can imagine we make the comma mandatory. FWIW, the Rust formatter always adds a comma (you need to enable this via an option, and in aptos-core, we have enabled this option). There is a merit to make the syntax canonical. Its really just not purpose of this PR to finalize such nitty-gritty details.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 18]
```diff
@@ -1413,6 +1456,55 @@ fn parse_for_loop(context: &mut Context) -> Result<(Exp, bool), Box<Diagnostic>>
     Ok((parsed_for_loop, ends_in_block))
 }
 
+// Match = "match" "(" <Exp> ")" "{" ( <MatchArm> ","? )*
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:1459]: Fixed (with the next push)

File: third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/variants/variants_ok.move:42]: If I would not have to read comments like "bad idea" on my PRs, responses would be also in nicer language, @brmataptos .

File: third_party/move/move-model/src/model.rs (hunk 5 above)
Comment (brmataptos) [reply third_party/move/move-model/src/model.rs:3467]: Oh, hadn't realized that this is in `StructEnv`.  I thought it was more global.

File: third_party/move/move-model/src/model.rs (hunk 6 above)
Comment (brmataptos) [reply third_party/move/move-model/src/model.rs:2330]: Yeah, I misunderstood the scope of get_fields_of_variant.

File: third_party/move/move-compiler-v2/tests/checking/specs/structs_err.exp [hunk 19]
```diff
@@ -1,6 +1,6 @@
 
 Diagnostics:
-error: field `xx` not declared in struct `S`
+error: field `xx` not declared in `S`
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/checking/specs/structs_err.exp:3]: It would be nice if we could at least clarify that this is a Struct or Variant.

**Review by brmataptos — APPROVED** [review]
Please change error message to "struct or variant" or add a TODO+issue to come back to it.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 20]
```diff
@@ -2460,11 +2551,15 @@ fn parse_address_specifier(context: &mut Context) -> Result<AddressSpecifier, Bo
 
 // Parse a struct definition:
 //      StructDecl =
-//          "struct" <StructDefName> ("has" <Ability> (, <Ability>)+)?
-//          ("{" Comma<FieldAnnot> "}" | ";")
+//          ( "struct" | "enum" )  <StructDefName> ("has" <Ability> (, <Ability>)+)?
+//          /*struct*/ ("{" Comma<FieldAnnot> "}" | ";")
+//          /*enum*/   "{" Comma<EnumVariant> "}"
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:2556]: Refactored the grammar a bit to make it clearer.

File: third_party/move/move-model/src/ty.rs [hunk 21]
```diff
@@ -2930,7 +2935,7 @@ impl TypeUnificationError {
                 let fields =
                     Self::print_fields(display_context.env, missing_fields.into_iter().cloned());
                 format!(
-                    "{} not declared in struct `{}`",
+                    "{} not declared in `{}`",
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/ty.rs:2938]: I fixed the issue, plus providing a hint now when selecting from an enum type that this is only allowed if the field is in all variants.

File: third_party/move/move-compiler-v2/tests/checking/specs/structs_err.exp (hunk 19 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/specs/structs_err.exp:3]: Done
